src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,658 @@
+ * The purpose is to replace standard library iostreams with something more usable, robust, and
+ * lightweight. One of the major issues with stdlib iostreams is the difficulty of obtaining useful
+ * error messages in a standardized way. The language spec allows for it, but implementations often
+ * map all error conditions to the same unhelpful message. std::ios::failure extends std::system_error
+ * but doesn't use errno as the error code. Also, they are heavyweight due to support for different
+ * locales (which is unnecessary in dinit).
+ *
+ * This library provides two classes for output and input: ostream and istream.
+ * See documentation for each class at the top of it.
+ *
+ * All functionality of this library is located in the "dio" namespace.
+ */
+constexpr unsigned IOSTREAM_BUFSIZE = 16384; // Size of the buffers
+
+namespace dio {
+
+/*
+ * Stream state bits:
+ *
+ * 1. "eofbit" indicates that the End Of File has been reached, and there is nothing left in
+ *    the buffer (istream only).
+ * 2. "buffer_failbit" indicates an attempt to use the buffer when the buffer pointer was nullptr
+ *    (e.g. failure to allocate a unique_ptr of streambuf) (ostream/istream).
+ * 3. "string_failbit" indicates a failed operation in std::string related calls (e.g. failure in
+ *    pushing a line from the buffer to the given std::string in get_line()) (istream only).
+ * 4. "io_failbit" indicates that a system I/O function failed, and the errno was set (ostream/istream).
+ *
+ * If all of the above bits are false, the stream status is "good".
+ *
+ * The current_state() function returns the stream's current status based on this table.
+ */
+enum io_states
+{
+    eofbit = 0x01,
+    buffer_failbit = 0x02,
+    string_failbit = 0x04,
+    io_failbit = 0x08
+};
+
+// Some empty classes for buffer operations implementation.
+// Used for accepting endl and flush in ostream::write() functions.
+class endline { };
+class flushbuf { };
+// Passing endl to those functions will put a '\n' and flush the stream's buffer.
+static constexpr endline endl;
+// Passing flush to those functions will flush the stream's buffer.
+static constexpr flushbuf flush;
+
+// Specialise system_error & runtime_error for more meaningful exceptions on state bits/vars:
+
+// on eofbit.
+class iostream_eof : public std::runtime_error
+{
+    public:
+    iostream_eof(const char *msg) : runtime_error(msg)
+    {
+    }
+};
+
+// on buffer_failbit, string_failbit.
+class iostream_internal_err : public std::runtime_error
+{
+    public:
+    iostream_internal_err(const char *msg) : runtime_error(msg)
+    {
+    }
+};
+
+// on io_failbit.
+class iostream_system_err : public std::system_error
+{
+    public:
+    iostream_system_err(const int error_code) : system_error(error_code, std::system_category())
+    {
+    }
+};
+
+using streambuf = cpbuffer<IOSTREAM_BUFSIZE>;
+
+// A simple type for keeping result of getc() functions.
+class getc_result
```

Comment:
Make this a `struct` as I said. `public:` can be removed. The constructor isn't needed.